shell/platform/android/io/flutter/embedding/engine/systemchannels/RestorationChannel.java

```diff
@@ -0,0 +1,153 @@
+ * Application before an Activity is created, this flag should be set to true because Android will
+ * only provide the restoration data to the Activity during the onCreate callback.
+ *
+ * <p>The current restoration data provided by the framework can be read via {@code
+ * getRestorationData()}.
+ */
+public class RestorationChannel {
+  private static final String TAG = "RestorationChannel";
+
+  public RestorationChannel(
+      @NonNull DartExecutor dartExecutor, @NonNull boolean waitForRestorationData) {
+    this(
+        new MethodChannel(dartExecutor, "flutter/restoration", StandardMethodCodec.INSTANCE),
+        waitForRestorationData);
+  }
+
+  RestorationChannel(MethodChannel channel, @NonNull boolean waitForRestorationData) {
+    this.channel = channel;
+    this.waitForRestorationData = waitForRestorationData;
+
+    channel.setMethodCallHandler(handler);
+  }
+
+  /**
+   * Whether the channel delays responding to the framework's initial request for restoration data
+   * until {@code setRestorationData} has been called.
+   *
+   * <p>If the engine never calls {@code setRestorationData} this flag must be set to false. If set
+   * to true, the engine must call {@code setRestorationData} either with the actual restoration
+   * data as argument or null if it turns out that there is no restoration data.
+   *
+   * <p>If the response to the framework's request for restoration data is not delayed until the
+   * data has been set via {@code setRestorationData}, the framework may intermittently initialize
+   * itself to default values until the restoration data has been made available. Setting this flag
+   * to true avoids that extra work.
+   */
+  public final boolean waitForRestorationData;
+
+  // Holds the the most current restoration data which may have been provided by the engine
+  // via "setRestorationData" or by the framework via the method channel. This is the data the
+  // framework should be restored to in case the app is terminated.
+  private byte[] restorationData;
+  private MethodChannel channel;
+  private MethodChannel.Result pendingFrameworkRequest;
+  private boolean engineHasProvidedData = false;
+  private boolean frameworkHasRequestedData = false;
+
+  /** Obtain the most current restoration data that the framework has provided. */
+  public byte[] getRestorationData() {
+    return restorationData;
+  }
+
+  /** Set the restoration data from which the framework will restore its state. */
+  public void setRestorationData(byte[] data) {
+    engineHasProvidedData = true;
+    if (pendingFrameworkRequest != null) {
+      // If their is a pending request from the framework, answer it.
+      pendingFrameworkRequest.success(data);
+      pendingFrameworkRequest = null;
+      restorationData = data;
+    } else if (frameworkHasRequestedData) {
+      // If the framework has previously received the engine's restoration data, push the new data
```

Comment:
I suppose this means the framework already asked for and have received a restoration data in the past (as opposed to the branch above). When should this happen in a non-add-to-app scenario? Or is this the use at-your-own-risk special case we talked about before? Maybe call that out in the comment for the impetuous future maintainer who might be thinking: "activity gets a bundle onCreate only once, let me delete this redundant handle".

Reply:
Provided more context.

Reply:
did you push? I don't see anything

Reply:
I can see it. Maybe reload? :P